velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp

```diff
@@ -542,8 +542,485 @@ struct ComplexTypeExtractor {
   }
 };
 
+template <typename V, typename Compare>
+struct MinMaxByNComplexCompareTypeAccumulator {
+  using TRawValue = typename RawValueExtractor<V>::TRawValue;
+  using BaseType =
+      MinMaxByNAccumulator<V, HashStringAllocator::Position, Compare>;
+  BaseType base;
+  ValueSet valueSet;
+
+  explicit MinMaxByNComplexCompareTypeAccumulator(
+      HashStringAllocator* allocator)
+      : base{allocator}, valueSet{allocator} {}
+
+  ~MinMaxByNComplexCompareTypeAccumulator() {
+    for (auto i = 0; i < base.heapValues.size(); ++i) {
+      freePair(base.heapValues[i]);
+    }
+  }
+
+  int64_t getN() const {
+    return base.n;
+  }
+
+  size_t size() const {
+    return base.size();
+  }
+
+  void checkAndSetN(DecodedVector& decodedN, vector_size_t row) {
+    return base.checkAndSetN(decodedN, row);
+  }
+
+  void compareAndAdd(
+      DecodedVector& decodedCompare,
+      vector_size_t compareIndex,
+      std::optional<V> value,
+      Compare& comparator) {
+    if (base.heapValues.size() < base.n) {
+      auto comparePosition = writeComplex(decodedCompare, compareIndex);
+      addToAccumulator(comparePosition, value, comparator);
+    } else {
+      const auto& topPair = base.heapValues.front();
+      if (comparator.compare(topPair, decodedCompare, compareIndex)) {
+        auto comparePosition = writeComplex(decodedCompare, compareIndex);
+        std::pop_heap(
+            base.heapValues.begin(), base.heapValues.end(), comparator);
+        base.heapValues.pop_back();
+        addToAccumulator(comparePosition, value, comparator);
+      }
+    }
+  }
+
+  /// Extract all values from 'heapValues' into 'rawValues' and 'rawValueNulls'
+  /// buffers. The heap remains unchanged after the call.
+  void extractValues(
+      FlatVector<V>& values,
+      vector_size_t offset,
+      Compare& comparator) {
+    std::sort_heap(base.heapValues.begin(), base.heapValues.end(), comparator);
```

Comment:
Looks like this code repeats a lot. Perhaps, it can be refactoring into a helper function that takes 'extractValue' lambda.

```
extractValues(base.heapValues, [&] (const auto& pair, auto i) {
   extractValue(pair, values, offset + i);
}
```

Reply:
Ditto extractPairs